### DISCUSSION TITLE
What do you want to see for v1.0?

### DISCUSSION THREAD

**Discussion by blacktm** [ruby2d/ruby2d#188] [Ideas]
Work has started on **Ruby 2D v1.0**! 🥳 What ideas do you have for new features, fixes, and improvements? Comment here and we'll discuss. **Disclaimer:** We may not be able to build every idea suggested. We'll do our best, but will consider the level of effort, if it's in line with the design aesthetic of the gem, and well, if we actually know how to do it (I actually don't know how to build _everything_, believe it or not, ha). If some features are big, they might be in point releases (e.g. v1.1) so we can still have a steady stream of new stuff to play with. Finally, [join the Discord server](https://discord.gg/QBWguEasV7) and the `#development` channel to chat about all this and more.

On the list (so far):
- Add a "direct draw" mode to the DSL, so you can draw to the window without creating objects, something like:
  ```ruby
  render do
    Square.draw(x: 10, y:20, color: 'red')
  end
  ```
- Support native builds on every platform (macOS, Windows, Linux, iOS, Android, HTML5/WebAssembly) using the [mruby](http://mruby.org) engine
- More window event handlers, for example:
  ```ruby
  on :window_close do
    puts 'Shutting down!'
  end
  ```
- Support the more classic game dev pattern, like:
  ```ruby
  class Window << Ruby2D::Window
    def draw; end
    def update; end
  end
  ```
- Generally clean up the DSL, like not showing objects by default when you call `.new`
- Start work on new new classes and capabilities, like cameras, particle systems, screen shakes, things like that
- Fix _all the bugs_ (we've got quite a few issues and pull requests open with those already)
- Add new shapes ([see here](https://github.com/ruby2d/ruby2d/discussions/188#discussioncomment-347035))
- Add GUI elements (e.g. buttons)

**Comment by Nakilon** [comment]
> the Discord server

You could have an IRC channel on Freenode to have more programmers than LoL and Minecraft players.

**Reply by realtradam** [reply]
I don't see why random LoL and MC players would join a programming server? Or why programmers wouldn't use Discord? Although if enough people did want an IRC channel it is possible to bridge the IRC channel and the Discord server using a bot to relay messages to both.

**Reply by blacktm** [reply]
Yea, they're definitely branching out more, and getting more popular with [open source projects](https://discord.com/open-source).

**Reply by Nakilon** [reply]
> Or why programmers wouldn't use Discord?

Why should they if they already have Freenode?
Discord became popular in the world only because the LoL developers were not proficient enough to implement the voice chat system in the game so they've cooperated with a **proprietary** single access point website platform with **clumsy interface** that **demands registration** to monetize integrations. And succeeded to do it because videogame players didn't have enough IT knowledge to know that there are Quakenet, Teamspeak, etc.

**Comment by UalaceCafe** [comment]
I have a few ideas based on my experience with other graphic/game frameworks.

1) I would love to see a native availability of border-only and borded shapes. I've implemented some for my own use, like a border-only (circumference) circle using the midpoint algorithm, and also things like quads with border and border-only quads. The thing is, it's too heavy drawing, for example, two quads or hundreds of circles (as we don't have a `pixel` shape) to draw the circumference of a circle.

2) Arcs and "curved" lines (e.g. Bezier and Splines).
For reference: https://p5js.org/reference/#/p5/arc
https://p5js.org/reference/#/p5/bezier

3) A P5js `beginShape` kind of block, where the user specify vertices inside this block and the gem automatically draws the shape according to these vertices.
For reference: https://p5js.org/reference/#/p5/beginShape

All of these are more graphical stuff, but I think they would be useful and cool features.

**Reply by blacktm** [reply]
Great ideas @UalaceCafe! Some of these definitely fall in the category of "stuff I don't know how to do," ha, but we could figure it out. Ideally we would implement in pure OpenGL, but another route could be drawing on a generic texture and rendering that. Kinda would be trading compute power for memory usage — costly to draw the first time by pixel, but then save it in a texture and just draw that each frame. Either way, might still need help getting the math right.

**Reply by UalaceCafe** [reply]
I could help with the math or logic, but someone else would have to do the programming hah
As for the future `Particle System` update, I would love to give it a hand, mine still works like a charm, but having it as part of the gem would be much better given the possibilities

**Reply by blacktm** [reply]
I'll take you up on that! Yes, would love to have a particle system in the gem itself. Never written one before, but I think I get the general idea (and how to make it fast on the GPU).

**Comment by realtradam** [comment]
I think having a built in physics engine would be a big boon for this project. Would it be feasible to incorporate something like Box2D?

> Start work on new new classes and capabilities, like cameras

I've actually started coding a solution for a camera system in Ruby2D. I wouldn't mind at all contributing to this project with that, however I'm curious if implementing this in ruby is fine, or if it would be preferred to be done in OpenGL/Simple2D/C? Personally I prefer doing it in Ruby but I do have some experience using OpenGL with C++ so perhaps I could help if a Ruby solution is non-ideal

**Reply by blacktm** [reply]
Yea, some basic 2D physics would be great to add. I [played around with Chipmunk](https://github.com/simple2d/chipmunk_demo/blob/master/chipmunk_demo.c) a bit ago, was pretty easy, fun to use. Box2D is more popular, but a C++ library (I wonder hard it is to use C++ in a Ruby extension... 🤔)

Re: cameras, oh yea, I remember you posting about that in Gitter. That's awesome, would also be great to add, thanks for the offer. I think generally we're trying to implement as much as possible in Ruby, only diving into C and OpenGL if it will really give it a performance boost. I often think Ruby is going to be slow, but it keeps surprising me, ha.

**Reply by realtradam** [reply]
> I often think Ruby is going to be slow, but it keeps surprising me, ha.

Thats great to hear! Ill keep working on the camera system in Ruby in that case.

I've never actually done making C(or C++?) extensions for ruby, but I've been very curious how to go about it. I only suggested Box2D because of how popular and complete it seems to be. I think Ill look into making C and C++ extensions once I finish working on the camera.  
I looked briefly and found quite a few(very old) Box2D wrapper gems on RubyGems so it should be possible?

**Reply by blacktm** [reply]
Sounds like a plan! Ruby _might_ be slow with heavy math, but if that’s the case, we can always try to offload some calculations to C. Oh, and we’re going to give Box2D a try (we’re starting to chat about it in Discord / #development).

**Comment by Maniczs** [comment]
Add a way to edit size attributes please.
For example, once a Text.new instance has been created, the size can no longer be edited. The only way would be by creating a new Text

Example:

`text1 = Text.new("blah",size: 50)`

`text1.size = 40`

`#runs code`
`#nothing happens`
`#text size still 50`

**Comment by davidgulyas** [comment]
It would be nice if text supported newlines.
Text.new("thing\nthing")
renders to 
thing<box indicating unsupported character>thing
